### PR TITLE
7047 - Added functionality to support the reverseQualityFilter flag

### PR DESCRIPTION
**Pull request by victor-chaparro**
Added functionality to support the reverseQualityFilter flag for receivers

Test Steps:
1. Submit a Full ELR HL7 using reports API - /api/reports?processing=async&default=aaa
2. Capture the Id of the submitted report
3. Verify Full ELR HL7 destination using report history API - /waters/report/{{reportId}}/history
4. Capture the receiver being routed
5. Update the reverseTheQualityFilter flag on this receiver inside organizations.yaml
6. Run ./gradlew reloadSettings to update settings for this receiver
7. Repeat steps 1 and 2
8. Verify the destination the previous destination is not present when calling the report history API

## Changes
- Added optional parameter reverseFilter to the evaluateFilterCondition
- Passing reverseTheQualityFilter flag when evaluating quality filters

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #7047

### CODE REVIEW

**Review by jimmyfagan — APPROVED** [review]
Looks good to me, just had one comment on the topic of interaction between `reverseFilter` and `defaultResponse` to consider, either way, marking approved.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 1]
```diff
@@ -243,7 +245,12 @@ class FHIRRouter(
             //           must have at least one of patient street, zip code, phone number, email
             //           must have at least one of order test date, specimen collection date/time, test result date
             passes = passes &&
-                evaluateFilterCondition(getQualityFilters(receiver, orgFilters), bundle, qualFilterDefaultResult)
+                evaluateFilterCondition(
+                    getQualityFilters(receiver, orgFilters),
+                    bundle,
+                    qualFilterDefaultResult,
+                    receiver.reverseTheQualityFilter
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:252]: You also need a unit test to test that the receiver's config is used correctly here.

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

**Comment by carlosfelix2** [comment]
Set the label of this PR to pipeline

**Comment by carlosfelix2** [comment]
The test steps section is not what you ran, but what you are telling others to run.  Obviously you ran this too.

**Comment by carlosfelix2** [comment]
The smoke tests are not really testing this new feature you added.  Perhaps add steps to change the reverse filter of a setting to manually test the flag works.  Hint: submit a full elr HL7, see what receiver is goes to, then change the flag in the DB or settings, send again, then see the data not go to that receiver.

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 2]
```diff
@@ -273,16 +281,19 @@ class FHIRRouter(
     internal fun evaluateFilterCondition(
```
Comment (TurkeyFried) [inline prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:281]: The doc for this method could be updated to explain the new param

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt [hunk 3]
```diff
@@ -273,16 +281,19 @@ class FHIRRouter(
...
         bundle: Bundle,
-        defaultResponse: Boolean
+        defaultResponse: Boolean,
+        reverseFilter: Boolean = false
     ): Boolean {
         // the filter needs to check all expressions passed in, or if the filter is null or empty it will return the
         // default response
-        return if (filter.isNullOrEmpty())
+        val result: Boolean = if (filter.isNullOrEmpty()) {
             defaultResponse
```
[the remark this answers is not in the document]
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:290]: Not the filters, the default _response_; the caller is presumably providing a default response to be "what do we do if there isn't a filter to use", so I don't think we should negate that.

**Review by TurkeyFried — CHANGES_REQUESTED** [review]
The code changes look good. My only comment is that the PR needs to either have manual testing steps or the code needs a Smoke (or Integration) test.

**Review by TurkeyFried — APPROVED** [review]
The updated test steps are good. I also saw this working on a pair earlier so thumbs up!

File: prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (jimmyfagan) [reply prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt:291]: Discussed this during standup today, and it sounds like the default response should be negated too, so this is doing what is wanted as it is; I'll close this conversation.

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt [hunk 4]
```diff
@@ -172,6 +170,27 @@ class DefaultFilterTests {
...
+        val one = Schema(name = "None", topic = "full-elr", elements = emptyList())
+        val metadata = Metadata(schema = one)
+
+        val engine = spyk(makeFhirEngine(metadata, settings, TaskAction.route) as FHIRRouter)
+        val bundle = FhirTranscoder.decode(fhirCodeP)
+
+        // act
+        val useDefaultResult = engine.evaluateFilterCondition(emptyList(), bundle, false, true)
+        // assert
+        assert(useDefaultResult)
```
Comment (carlosfelix2) [inline prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt:186]: Use assertThat instead in all your tests.  This is the program directed unit test framework. Assert is now only used if you cannot use assertThat which is rare.  See assertk

**Review by carlosfelix2 — CHANGES_REQUESTED** [review]

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt (hunk 4 above)
Comment (jimmyfagan) [reply prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt:186]: Is that requirement meant to improve readability or is there some other reason? Is there any kind of document or decision record that we could reference to have a better idea of these kinds of standards, or is it more of an informal shared understanding at this point?

Comment (victor-chaparro) [reply prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt]: Updated unit tests to user assertThat. Ran unit tests after making these changes and they're all passing.

File: prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt [hunk 5]
```diff
@@ -170,7 +171,28 @@ class DefaultFilterTests {
...
+        val engine = spyk(makeFhirEngine(metadata, settings, TaskAction.route) as FHIRRouter)
+        val bundle = FhirTranscoder.decode(fhirCodeP)
+
+        // act
+        val useDefaultResult = engine.evaluateFilterCondition(emptyList(), bundle, false, true)
+        // assert
+        assertThat(useDefaultResult).isTrue()
+
+        // act
+        val useBundleResult = engine.evaluateFilterCondition(engine.processingModeFilterDefault, bundle, false, false)
```
Comment (carlosfelix2) [inline prime-router/src/test/kotlin/fhirengine/engine/fhirRouterTests/DefaultFilterTests.kt:193]: Not asking you to change this right now, but there is no need for the useBundleResult val.  Just put the code inside the assertThat().  Same for all the other below.  I know you did not write those.

File: prime-router/src/main/kotlin/Receiver.kt [hunk 6]
```diff
@@ -87,7 +87,8 @@ open class Receiver(
         jurisdictionalFilter: ReportStreamFilter = emptyList(),
         qualityFilter: ReportStreamFilter = emptyList(),
         routingFilter: ReportStreamFilter = emptyList(),
-        processingModeFilter: ReportStreamFilter = emptyList()
+        processingModeFilter: ReportStreamFilter = emptyList(),
+        reverseTheQualityFilter: Boolean = false
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/Receiver.kt:91]: Update the organizations.yml to add `reverseTheQualityFilter: false` to the "full-elr-hl7" receiver

**Review by carlosfelix2 — APPROVED** [review]

**Pull request merged**